runtime/common/ArgumentConversion.cpp

```diff
@@ -99,12 +99,245 @@ static Value genConstant(OpBuilder &, cudaq::cc::StructType, void *,
 static Value genConstant(OpBuilder &, cudaq::cc::ArrayType, void *,
                          ModuleOp substMod, llvm::DataLayout &);
 
+/// Create callee.init_N that initializes the state
+/// Callee:
+/// func.func @__nvqpp__mlirgen__callee(%arg0: i64) {
+///   %0 = cc.alloca i64
+///   cc.store %arg0, %0 : !cc.ptr<i64>
+///   %1 = cc.load %0 : !cc.ptr<i64>
+///   %2 = quake.alloca !quake.veq<?>[%1 : i64]
+///   %3 = quake.extract_ref %2[1] : (!quake.veq<?>) -> !quake.ref
+///   quake.x %3 : (!quake.ref) -> ()
+///   return
+/// }
+/// callee.init_N:
+/// func.func private @callee.init_0(%arg0: !quake.veq<?>, %arg0: i64) ->
+/// !!quake.veq<?> {
+///   %1 = quake.extract_ref %arg0[1] : (!quake.veq<2>) -> !quake.ref
+///   quake.x %1 : (f64, !quake.ref) -> ()
+///   return %arg0: !quake.veq<?>
+/// }
+static void createInitFunc(OpBuilder &builder, ModuleOp sourceMod,
+                           func::FuncOp calleeFunc,
+                           std::string &initKernelName) {
+  OpBuilder::InsertionGuard guard(builder);
+  builder.setInsertionPointToEnd(sourceMod.getBody());
```

Comment:
Are we creating _argument substitution_ information in the substitution module here?

It should not be assumed that the source module can be modified in place or that it is going to be the same module as the final result, after argument synthesis.

So, we have 3 modules:
   1. original source (assume it is const)
   2. a substitution module full of substitution patterns
   3. a result module (to be generated by a different pass)

Reply:
Changed to adding new functions in the parent substitution module (this works after making the synthesis transitive in https://github.com/NVIDIA/cuda-quantum/pull/2644